boa/src/builtins/string/mod.rs

```diff
@@ -36,46 +36,52 @@ use std::{
 pub(crate) struct String;
 
 impl String {
+    fn this_string_value(this: &Value, ctx: &mut Interpreter) -> Result<StdString, Value> {
+        match this.data() {
+            ValueData::String(ref string) => return Ok(string.clone()),
+            ValueData::Object(ref object) => {
+                let object = object.borrow();
+                if let Some(string) = object.as_string() {
+                    return Ok(string.clone());
+                }
+            }
+            _ => {}
+        }
+
+        ctx.throw_type_error("'this' is not a string")?;
+        unreachable!();
+    }
+
     /// [[Construct]] - Creates a new instance `this`
     ///
     /// [[Call]] - Returns a new native `string`
     /// <https://tc39.es/ecma262/#sec-string-constructor-string-value>
     pub(crate) fn make_string(
         this: &mut Value,
         args: &[Value],
-        _: &mut Interpreter,
+        ctx: &mut Interpreter,
     ) -> ResultValue {
         // This value is used by console.log and other routines to match Obexpecty"failed to parse argument for String method"pe
         // to its Javascript Identifier (global constructor method name)
-        let s = args.get(0).unwrap_or(&Value::string("")).clone();
-        let length_str = s.to_string().chars().count();
-
-        this.set_field("length", Value::from(length_str as i32));
+        let string = match args.get(0) {
+            Some(ref value) => ctx.to_string(value)?,
+            None => StdString::new(),
+        };
```

Comment:
```suggestion
        let string = match args.get(0).map(|value| ctx.to_string(value)).unwrap_or_else(|| Ok(StdString::new()))?;
```

Reply:
we can't use the `?` try operator in lambdas.

Reply:
But in this case, the `?` operator would be outside of the lambda, right?

Reply:
> But in this case, the `?` operator would be outside of the lambda, right?

I guess so, but i think it becomes less readable in this case, having all the logic in one line.